### PR TITLE
create a shared library to make creating wallets easier.

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

**Comment by sickpig** [comment]
is this ready to be reviewed?

**Comment by gandrewstone** [comment]
Yes.  Also, I'd appreciate some help from windows guys on the compilation error in that platform.

**Comment by ptschip** [comment]
Getting this error on win64 when i compile:

                 from script/sigcommon.cpp:5:
d:/Deps/x64/boost_1_61_0/boost/asio/detail/socket_types.hpp:24:4: error: #error WinSock.h has already been included
 "#  error WinSock.h has already been included"
    ^

**Comment by ptschip** [comment]
Not sure what the sigcommon.cpp is there for?  most of it is commented out and it doesn't have a header file nor any Copywrite header.

**Comment by ptschip** [comment]
Just to document it, this looks like where the error message is getting triggered in boost socket_types.hpp:

"
#include <boost/asio/detail/config.hpp>

#if defined(BOOST_ASIO_WINDOWS_RUNTIME)
// Empty.
#elif defined(BOOST_ASIO_WINDOWS) || defined(__CYGWIN__)
# if defined(_WINSOCKAPI_) && !defined(_WINSOCK2API_)
#  error WinSock.h has already been included
# endif // defined(_WINSOCKAPI_) && !defined(_WINSOCK2API_)
"

**Comment by gandrewstone** [comment]
sigcommon.cpp is a file that contains routines that are common to the new shared library and also used in bitcoin.  It doesn't need a header because the routines are already defined in their original headers.  I expect that as this library grows in functionality, more stuff will be put into it.

The .so needs to have a simple C interface because a lot of higher level (interpreted) languages have the ability to load up a .so and call C functions.  This is the intention of libbitcoincash.so, and so it contains a lot of wrapper functions that present code that is C++ under the hood as normal C.  But compared to C++, this API is pretty awkward.  So it does not make sense for bitcoind to link with libbitcoincash.so.

WRT the error, I'm not doing anything fancy/weird with boost.  So its weird that this code would get an error...

**Comment by ptschip** [comment]
@gandrewstone you're missing "compat.h" in both sigcommon.cpp and cashlib.cpp

**Comment by ptschip** [comment]
this is what is important in compat.h...winsock2.h has to come first.

#include <winsock2.h> // Must be included before mswsock.h and windows.h

#include <mswsock.h>
#include <windows.h>
#include <ws2tcpip.h>

File: src/cashlib/cashlib.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
+#include <openssl/rand.h>
```
Comment (ptschip) [inline src/cashlib/cashlib.cpp]: can you add the copywrite to both cashlib and sigcommon

+# Copyright (c) 2015-2018 The Bitcoin Unlimited developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.

**Comment by gandrewstone** [comment]
New windows problem:
.libs/libbitcoincash_la-utiltime.o:utiltime.cpp:(.text+0x460): undefined reference to `_imp__clock_gettime'

File: qa/rpc-tests/test_framework/cashlib/cashlib.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+def hash256(s):
+    """Return the double SHA256 hash (what bitcoin typically uses) of the passed binary data
+
+    >>> hexlify(hash256("There was a terrible ghastly silence".encode()))
+    b'730ac30b1e7f4061346277ab639d7a68c6686aeba4cc63280968b903024a0a40'
+    """
+    return sha256(sha256(s))
+
+def hash160(msg):
+    """RIPEME160(SHA256(msg)) -> bytes"""
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/cashlib/cashlib.py]: s/RIPEME/RIPEMD/

File: qa/rpc-tests/test_framework/cashlib/cashlib.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+from ctypes import *
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/cashlib/cashlib.py:4]: missing copyright headers

File: qa/rpc-tests/test_framework/cashlib/cashlib.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            dir_path = os.path.dirname(os.path.realpath(__file__))
+            cashlib = CDLL(dir_path + os.sep + libbitcoincashfile)
+    else:
+        cashlib = CDLL(libbitcoincashfile)
+    if cashlib is None:
+        raise Error("Cannot find %s shared library", libbitcoincashfile)
+
+
+# Serialization/deserialization tools
+def sha256(s):
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/cashlib/cashlib.py:42]: this function is copy/pasted from `nodemessages.py`, use the on already define rather than add code duplication.

File: qa/rpc-tests/test_framework/cashlib/cashlib.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    """Return the sha256 hash of the passed binary data
+
+    >>> hexlify(sha256("e hat eye pie plus one is O".encode()))
+    b'c5b94099f454a3807377724eb99a33fbe9cb5813006cadc03e862a89d410eaf0'
+    """
+    return hashlib.new('sha256', s).digest()
+
+
+def hash256(s):
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/cashlib/cashlib.py:51]: ditto

File: qa/rpc-tests/test_framework/cashlib/cashlib.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    """Return the double SHA256 hash (what bitcoin typically uses) of the passed binary data
+
+    >>> hexlify(hash256("There was a terrible ghastly silence".encode()))
+    b'730ac30b1e7f4061346277ab639d7a68c6686aeba4cc63280968b903024a0a40'
+    """
+    return sha256(sha256(s))
+
+def hash160(msg):
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/cashlib/cashlib.py]: ditto

File: src/Makefile.am [hunk 7]
```diff
@@ -463,6 +495,22 @@ bitcoin_tx_LDADD = \
 bitcoin_tx_LDADD += $(BOOST_LIBS) $(CRYPTO_LIBS)
 #
 
+# bitcoinconsensus library #
```
Comment (sickpig) [inline src/Makefile.am]: change the comment to bitcoincash library

File: src/cashlib/cashlib.cpp [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+int LogPrintStr(const std::string &str) { return str.size(); }
+namespace Logging
+{
+uint64_t categoriesEnabled = 0; // 64 bit log id mask.
+};
+
+// helper functions
+namespace
+{
+#if 0
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp]: remove disabled code.

File: src/cashlib/cashlib.cpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+CKey LoadKey(unsigned char *src)
+{
+    CKey secret;
+    secret.Set(src, src + 32, true);
+    return secret;
+}
+
+// From core_read.cpp #include "core_io.h"
+bool DecodeHexTx(CTransaction &tx, const std::string &strHexTx)
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp:39]: why redefining `DecodeHexTx`? is already defined in ` core_read.cpp ` (as you noted already in the comment. doesn't his introduce code duplication?

File: src/cashlib/cashlib.cpp [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    catch (const std::exception &)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+#if 0
+bool DecodeHexTx(CTransaction &tx, const char* cHexTx)
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp]: remove disabled code

File: src/cashlib/cashlib.cpp [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        unsigned char val = data[i];
+        ret.push_back(hexxlat[val >> 4]);
+        ret.push_back(hexxlat[val & 15]);
+    }
+    return ret;
+}
+}
+
+#if 0
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp]: remove disabled code

File: src/cashlib/cashlib.cpp [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+    std::string s = GetHex(val, length);
+    if (s.size() >= resultLen)
+        return 0; // need 1 more for /0
+    strncpy(result, s.c_str(), resultLen);
+    return 1;
+}
+
+
+/** Return random bytes from cryptographically acceptable random sources */
+extern "C" int RandomBytes(unsigned char *buf, int num)
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp:86]: why not using `GetRandBytes` defined in `random.cpp`

File: src/httpserver.cpp [hunk 13]
```diff
@@ -13,6 +13,7 @@
 #include "ui_interface.h"
 #include "util.h"
 
+#include <mutex>
```
Comment (sickpig) [inline src/httpserver.cpp:16]: is seems unrelated. why we need it?

File: src/script/sigcommon.cpp [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+{
+    CHashWriter ss(SER_GETHASH, 0);
+    for (unsigned int n = 0; n < txTo.vout.size(); n++)
+    {
+        ss << txTo.vout[n];
+    }
+    return ss.GetHash();
+}
+
+#if 0 // Not used right now
```
Comment (sickpig) [inline src/script/sigcommon.cpp]: remove dead code

File: src/script/sigcommon.cpp [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+{
+    CKey secret;
+    unsigned char *src = val.begin();
+    secret.Set(src, src+32, true);
+    return secret;
+}
+#endif
+
+#if 0
+// From core_read.cpp #include "core_io.h"
```
Comment (sickpig) [inline src/script/sigcommon.cpp]: remove dead code

File: src/script/sigcommon.cpp [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+
+        uint256 sighash = ss.GetHash();
+        // printf("SigHash: %s\n", sighash.GetHex().c_str());
+        return sighash;
+    }
+    return one;
+}
+
+
+#if 0
```
Comment (sickpig) [inline src/script/sigcommon.cpp]: remove dead code

File: src/script/sigcommon.cpp [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+
+    if (!key.Sign(hash, sig))
+        return std::string();
+    return GetHex(sig);
+}
+#endif
+
+
+#if 0
+std::string SignTx(std::string txHex, unsigned int inputIdx, int64_t inputAmount, uint32_t nHashType, std::string hexkey)
```
Comment (sickpig) [inline src/script/sigcommon.cpp]: remove dead code

File: src/script/sigcommon.cpp [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+    sig.push_back((unsigned char)nHashType);
+    std::string s = GetHex(sig);
+    if (s.size() >= resultLen) return 0;
+    strncpy(result, s.c_str(), resultLen);
+    return 1;
+}
+#endif
+
+#if 0
+bool static IsValidSignatureEncoding(const std::vector<unsigned char> &sig)
```
Comment (sickpig) [inline src/script/sigcommon.cpp]: ditto

**Review by sickpig — COMMENTED** [review]
mainly noticing code duplication and a bunch of disabled code that should be removed.

**Comment by awemany** [comment]
Do I understand it correctly that functions such as the Bin2Hex function that returns into a caller-supplied char* array are like that just so that it is easier to import it from python using the ctypes interface?

I am wondering whether it would be a good idea to somehow mark those functions in some way to be meant 'just for python export/import' so that no one starts to use them somewhere else in the code. What do you think?

Because, as much as I can see how it makes life easy regarding py-interfacing, in a more general environment, that kind of 'fill in char* array' seems ugly and dangerous.

File: qa/rpc-tests/test_framework/cashlib/cashlib.py (hunk 4 above)
Comment (gandrewstone) [reply qa/rpc-tests/test_framework/cashlib/cashlib.py:42]: cashlib is supposed to stand alone, not be tightly integrated into the test_framework.  These APIs will likely be replaced with their C versions so that they can be accessed from other languages, and be high performance (although python likely uses a c implementation of hashlib, its probably not optimized assembly like bitcoin has).  Right now I implemented them in python via copy-paste so that I can gauge their usefulness as part of the cashlib API.

File: src/cashlib/cashlib.cpp (hunk 9 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp:39]: cashlib.cpp only becomes libbitcoincash.so which is not linked into bitcoind.  There were other symbols in core_read.cpp that was causing libbitcoincash.so to pull in other .cpp recursively so I had to pull this out.

File: src/cashlib/cashlib.cpp (hunk 10 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp]: The APIs provided in this library are useful which is why I issued the PR.  But are there other useful APIs?  While I was experimenting with APIs I wrote this code, and I'm not convinced that it won't be useful, although it isn't used now.

"cashlib" is not done.  New APIs will be useful, and other languages may need different formulations of the same API.  For example, passing data as hex strings rather then binary.  

But if we wait to PR until it is "done", it will be too large, and require constant merging.  I would like to merge cashlib in now as "experimental" (APIs may change), and continue to work on it with a series of smaller PRs.

File: src/httpserver.cpp (hunk 13 above)
Comment (gandrewstone) [reply src/httpserver.cpp:16]: windows build was giving me an error std::mutex undefined.  IDK why travis was not seeing it.

**Comment by gandrewstone** [comment]
@awemany, yes, using C APIs only makes wrapping in some interpreted languages easy.  I think that its not just python, although I plan to discover more as I add wrappers for more languages.  

Don't worry about accidental use.  You can't use anything in cashlib.cpp in any other bitcoin program, you will get undefined symbols.  Its only linked to libbitcoincash.so which is not linked to any of our executables.

File: src/cashlib/cashlib.cpp (hunk 12 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp:86]: It logs and asserts.  Also it is name mangled, so would have to wrap anyway.  This routine returns the number of random bytes generated, so that result can be checked and a higher level language exception can be thrown.

File: src/cashlib/cashlib.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp]: done

**Comment by gandrewstone** [comment]
OK I think I addressed all comments, removed dead code, etc.

**Comment by sickpig** [comment]
`cashlibtest.py` failed 

```
2017-08-02 10:46:40 Acceptable block: ver:20000000 t
stderr:
   File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 191, in main
    self.run_test()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/cashlibtest.py", line 102, in run_test
    tx2id = self.nodes[0].sendrawtransaction(hexlify(tx2.serialize()).decode("utf-8"))
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/authproxy.py", line 166, in __call__
    raise JSONRPCException(response['error'])
Traceback (most recent call last):
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/cashlibtest.py", line 118, in <module>
    MyTest().main()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 239, in main
    print("".join(deque(open(f), MAX_LINES_TO_PRINT)))
KeyboardInterrupt
```

seems the usual case when travis freaks out. 

restarting the Job.

File: src/script/sigcommon.cpp [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    unsigned int nIn,
+    uint32_t nHashType,
+    const CAmount &amount,
+    size_t *nHashedOut)
+{
+    // printf("sighash: scriptSize %d, nIn %d, nHashType %d, amount %lld\n", scriptCode.size(), nIn, nHashType, amount);
+    // for(unsigned int i=0;i<scriptCode.size(); i++)
+    //    printf("%02x",scriptCode[i]);
+    // printf("\n");
+    static const uint256 one(uint256S("0000000000000000000000000000000000000000000000000000000000000001"));
```
Comment (ptschip) [inline src/script/sigcommon.cpp]: seem to be old comments used for testing the code that can be taken out.

File: qa/rpc-tests/test_framework/cashlib/cashlib.py (hunk 3 above)
Comment (gandrewstone) [reply qa/rpc-tests/test_framework/cashlib/cashlib.py:4]: done

File: qa/rpc-tests/cashlibtest.py [hunk 20]
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+# Copyright (c) 2015-2018 The Bitcoin Unlimited developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+# This is a template to make creating new QA tests easy.
+# You can also use this template to quickly start and connect a few regtest nodes.
+
```
Comment (ptschip) [inline qa/rpc-tests/cashlibtest.py]: "This is a template..."  should take out , maybe add something else or nothing at all?

File: qa/rpc-tests/cashlibtest.py [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+        sig2 = cashlib.signTxInput(tx2, 0, amt, output, destPrivKey, sighashtype)
+        tx2.vin[0].scriptSig = cashlib.spendscript(sig2, destPubKey)
+
+        tx2id = self.nodes[0].sendrawtransaction(hexlify(tx2.serialize()).decode("utf-8"))
+
+        # Check that all tx were created, and commit them
+        assert self.nodes[0].getmempoolinfo()["size"] == 2
+        self.nodes[0].generate(1)
+
+
```
Comment (ptschip) [inline qa/rpc-tests/cashlibtest.py]: I would do one last check here to make sure the block was propagated correctly and that the mempool is empty.

+ self.sync_all()
+ assert self.nodes[0].getmempoolinfo()["size"] == 0

File: qa/rpc-tests/cashlibtest.py (hunk 20 above)
Comment (gandrewstone) [reply qa/rpc-tests/cashlibtest.py]: done

File: qa/rpc-tests/cashlibtest.py (hunk 21 above)
Comment (gandrewstone) [reply qa/rpc-tests/cashlibtest.py]: I did a sync_blocks and check.  We should use sync_all only when absolutely necessary because mempools may not ever get in sync if the bloom filter has a false positive.  This makes the test fail (very rarely, depending on the txid) forcing us to rekick travis.

**Review by ptschip — APPROVED** [review]
